tools/@aws-cdk/node-bundle/src/bundle.ts

```diff
@@ -0,0 +1,311 @@
+    // useless, so lets generate it of the bat.
+    const { dependencies, output } = this.esbuild();
+    this.dependencies = dependencies;
+    this.output = output;
+  }
+
+  /**
+   * Validate the state of the project with respect to bundling.
+   *
+   * This method will validate both circular imports and notice file attributions.
+   * To validate only one or the other, use the `validateNotice` and `validateCircularImports`.
+   *
+   * It never throws an exception, instead it returns a report of violations. The Caller is responsible
+   * for inspecting those violations and act accordingaly.
+   *
+   * If no violations are found, the return value will be undefined.
+   */
+  public validate(): BundleViolations | undefined {
+    const importsViolations = this.validateCircularImports();
+    const noticeViolations = this.notice.validate();
+
+    if (!importsViolations && !noticeViolations) {
+      return undefined;
+    }
+
+    return { notice: noticeViolations, imports: importsViolations };
+  }
+
+  /**
+   * Validate the package (and its dependencies) don't exhibit circular imports.
+   *
+   * It never throws an exception, instead it return a rerpot of violations. The Caller is responsible
+   * for inspecting those violations and act accordingaly.
+   *
+   * If no violations are found, the return value will be undefined
+   */
+  public validateCircularImports(): CircularImportsViolations | undefined {
+    console.log('Validating circular imports');
+    const packages = [this.packageDir, ...this.dependencies.map(d => d.path)];
+    try {
+      // we don't use the programatic API since it only offers an async API.
+      // prefer to stay sync for now since its easier to integrate with other tooling.
+      // will offer an async API further down the road.
+      shell(`${require.resolve('madge/bin/cli.js')} --warning --no-color --no-spinner --circular --extensions js ${packages.join(' ')}`, { quiet: true });
+      return undefined;
+    } catch (e: any) {
+      return { summary: e.stdout.toString() };
+    }
+  }
+
+  public validateNotice(): NoticeViolations | undefined {
+    console.log('Validating notice');
+    return this.notice.validate();
+  }
+
+  public createNotice() {
+    this.notice.create(this.copyright);
+  }
+
+  public fix() {
+    console.log('Generating notice file');
+    this.notice.create(this.copyright);
+  }
+
+  public pack(target: string) {
+
+    // double check, make sure we don't package something invalid.
+    const violations = this.validate();
+    if (violations) {
+      throw new Error('Unable to pack due to validation errors. Please run validate() to inspect them and fix.');
+    }
+
+    console.log('Creating package');
+
+    if (!fs.existsSync(target)) {
+      console.log(`✖ Target doesnt exist: ${target}`);
+      process.exit(1);
+    }
+
+    if (!fs.lstatSync(target).isDirectory()) {
+      console.log(`✖ Target must be a directory: ${target}`);
```

Comment:
Is there any reason why it's important that it's not a symlink to a directory?

Reply:
Assuming `fs.coySync` follows symlinks, nope. I'll check and change 👍